opm/simulators/linalg/bda/rocsparseWellContributions.cpp

```diff
@@ -0,0 +1,239 @@
+    }
+
+namespace Opm
+{
+
+#ifdef __HIP__
+/// HIP kernel to apply the standard wellcontributions
+__global__ void stdwell_apply(
+            const double *Cnnzs,
+            const double *Dnnzs,
+            const double *Bnnzs,
+            const unsigned *Ccols,
+            const unsigned *Bcols,
+            const double *x,
+            double *y,
+            const unsigned dim,
+            const unsigned dim_wells,
+            const unsigned *val_pointers)
+{
+    unsigned wgId = blockIdx.x;
+    unsigned wiId = threadIdx.x;
+    unsigned valSize = val_pointers[wgId + 1] - val_pointers[wgId];
+    unsigned valsPerBlock = dim*dim_wells;
+    unsigned numActiveWorkItems = (blockDim.x/valsPerBlock)*valsPerBlock;
+    unsigned numBlocksPerWarp = blockDim.x/valsPerBlock;
+    unsigned c = wiId % dim;
+    unsigned r = (wiId/dim) % dim_wells;
+    double temp;
+
+    extern __shared__ double localSum[];
+    double *z1 = localSum + gridDim.x;
+    double *z2 = z1 + dim_wells;
+
+    localSum[wiId] = 0;
+    if(wiId < numActiveWorkItems){
+        unsigned b = wiId/valsPerBlock + val_pointers[wgId];
+        while(b < valSize + val_pointers[wgId]){
+            int colIdx = Bcols[b];
+            localSum[wiId] += Bnnzs[b*dim*dim_wells + r*dim + c]*x[colIdx*dim + c];
+            b += numBlocksPerWarp;
+        }
+
+        // merge all blocks in this workgroup into 1 block
+        // if numBlocksPerWarp >= 3, should use loop
+        // block 1:     block 2:
+        //  0  1  2     12 13 14
+        //  3  4  5     15 16 17
+        //  6  7  8     18 19 20
+        //  9 10 11     21 22 23
+        // workitem i will hold the sum of workitems i and i + valsPerBlock
+        if(wiId < valsPerBlock){
+            for (unsigned i = 1; i < numBlocksPerWarp; ++i) {
+                localSum[wiId] += localSum[wiId + i*valsPerBlock];
+	    }
+        }
+
+        if(c == 0 && wiId < valsPerBlock){
+            for(unsigned i = dim - 1; i > 0; --i){
+                localSum[wiId] += localSum[wiId + i];
+            }
+            z1[r] = localSum[wiId];
+        }
+    }
+
+    __syncthreads();
+
+    if(wiId < dim_wells){
+        temp = 0.0;
+        for(unsigned i = 0; i < dim_wells; ++i){
+            temp += Dnnzs[wgId*dim_wells*dim_wells + wiId*dim_wells + i]*z1[i];
+        }
+        z2[wiId] = temp;
+    }
+
+    __syncthreads();
+
+    if(wiId < dim*valSize){
+        temp = 0.0;
+        unsigned bb = wiId/dim + val_pointers[wgId];
+        for (unsigned j = 0; j < dim_wells; ++j){
+            temp += Cnnzs[bb*dim*dim_wells + j*dim + c]*z2[j];
+        }
+
+        int colIdx = Ccols[bb];
+        y[colIdx*dim + c] -= temp;
+    }
+}
+#endif
+
+
+void WellContributionsRocsparse::apply_stdwells([[maybe_unused]] double *d_x,
+                                                [[maybe_unused]] double *d_y){
+#ifdef __HIP__
+    unsigned gridDim = num_std_wells;
+    unsigned blockDim = 32;
```

Comment:
Indeed, 32 was NVIDIA specific while AMD used 64 until the RDNA microarchitecture appeared, which dropped that to 32 (although 64 is still possible). However, 64 is still the default for the other microarchitecture CDNA used by the AMD Instinct MI210 GPU that we are using for our experiments, so I changed this to 64 in the code. I searched for a bit to find out if this can be queried at runtime but I did not find anything useful.